### DISCUSSION TITLE
Feature Request: Enhanced Responsive Modal Widths

### DISCUSSION THREAD

**Discussion by andrewdwallo** [filamentphp/filament#12367] [Ideas]
## Summary
Introduce a new method, `responsiveModalWidth()`, to Filament to complement the existing `modalWidth()` by providing responsive behavior for modals based on TailwindCSS breakpoints. This method allows developers to specify an alternative width for smaller screens.

## Feature Description
The `responsiveModalWidth()` method would allow for dynamic adjustment of modal widths based on device screen size:
- **modalWidth:** The width of the modal that should apply when the breakpoint condition is met.
- **breakpoint:** The TailwindCSS breakpoint at which the modalWidth should apply.

## Example Usage
```php
use Filament\Support\Enums\MaxWidth;
 
Action::make('deleteBankConnection')
    ->form([
        // ...
    ])
    ->action(function (array $data): void {
        // ...
    })
    ->modalWidth(MaxWidth::TwoExtraLarge)
    ->responsiveModalWidth(MaxWidth::Screen, 'sm')
```

## Example Result

https://github.com/filamentphp/filament/assets/104294090/3695dbc1-a3fc-4b78-ad05-941b388d9c08

### Improved User Experience on Mobile Devices

By enabling a specific modal width for smaller screens, such as full-screen modals on mobile devices, the `responsiveModalWidth()` method enhances user experience by making better use of limited space. This adaptation is crucial for maintaining usability and accessibility, as users on mobile devices can interact with modal content more comfortably without feeling cramped.

### Consistency in Responsive Design

The method supports the best practices in responsive design by allowing modals to adapt based on predefined breakpoints of TailwindCSS. This consistency ensures that all users, regardless of their device type, encounter modals that are appropriately sized for their screens, which enhances the overall user interface.

### Simplifying Development

This approach simplifies the development process by reducing the complexity typically involved in manually managing different modal sizes for various screen sizes. Developers can set a universal modal size with an exception for smaller screens in a concise, manageable way, which streamlines both development and maintenance.

### Summary of Benefits
This method significantly enhances the capability of Filament modals to respond to different screen sizes, providing a powerful tool for developers to create adaptive and user-friendly interfaces. This functionality not only meets the needs of a diverse user base but also aligns with modern web development practices, emphasizing flexibility, consistency, and scalability.

## My Current Approach
Because I am currently limited to only being able to specify the modal width using the `modalWidth()` method, I utilize the following to be able to achieve something similar to the video example I showed above.

### Component Class
```php
class ListInstitutions extends Component implements HasActions, HasForms
{
    use InteractsWithActions;
    use InteractsWithForms;

    public string $modalWidth;
    
    public function deleteBankConnection(): Action
    {
        return Action::make('deleteBankConnection')
            ->modalWidth(fn () => $this->modalWidth)
            ->modalFooterActionsAlignment(fn () => $this->modalWidth === 'screen' ? Alignment::Center : Alignment::Start)
            
```

### Blade view
```blade
@script
<script>
    const mobileSize = window.matchMedia("(max-width: 480px)");

    let data = Alpine.reactive({ windowWidth: 'max-w-2xl' });

    // Add a media query change listener
    mobileSize.addEventListener('change', (e) => {
        data.windowWidth = e.matches ? 'screen' : 'max-w-2xl';
    });

    Alpine.effect(() => {
        $wire.$set('modalWidth', data.windowWidth);
    });
</script>
@endscript
```

This could be an issue when using something such as Filament Resources though. @danharrin I would be interested in your input if this is something you could see being beneficial for developers. I was going to try and make the PR myself but thought it would be best to discuss if it was suitable for Filament beforehand to avoid wasting any time. If it is deemed suitable I would also like to discuss different approaches towards implementing this feature and whatever would be best for Filament. 

**I would also like to hear anyone else's thoughts on this.**

**Comment by andrewdwallo** [comment]
@zepfietje It would also be great to hear your opinion on this considering that you are the lead UI designer of Filament and a UI master.

**Comment by zepfietje** [comment]
@andrewdwallo, could you provide some use cases where custom modal width is required for different breakpoints? I'm not sure if this is a thing common enough to maintain the added complexity in the framework.

E.g. are you only using this to make the modal full-screen on mobile?

**Reply by andrewdwallo** [reply]
Yes, I am currently only using it to make the modal Fullscreen on mobile. If you think that would be the only use case, then that would definitely be simpler to maintain I suppose. It would make sense to really only need that instead of having different modal widths at different breakpoints other than for mobile.

**Reply by danharrin** [reply]
To be honest, I've thought about making modals full screen on mobile by default anyway, instead of the in-between we have now with the border of content around the edge

**Reply by andrewdwallo** [reply]
That would make sense. I also believe fullscreen modal width for mobile would really be the only thing that would look best.

**Reply by andrewdwallo** [reply]
As well as having the modal footer actions stacked when having fullscreen modal on mobile like the example I showed above.

**Reply by danharrin** [reply]
can you PR that change instead then, on <sm devices?

**Reply by andrewdwallo** [reply]
Yeah, I can start to when I get home after work. Is there any specific way you want me to go about implementing it? Just want to make sure beforehand. Just any details that you think might be important.

**Reply by danharrin** [reply]
After a chat with @zepfietje it seems he disagrees, I'll let him explain here, so please don't start yet

**Reply by zepfietje** [reply]
Yeah, to start with I think this would be too big of a change to introduce in a minor version.

The main issue is that full-screen modals on mobile aren't more accessible in all cases. I do understand it makes sense when there's a lot of modal content, but that's just a portion of modal usage. Imagine a modal that simply requests a name and email address, or even a modal with a single form field. Covering the entire page would hide all context for the user, and it would degrade the user experience compared to having a modal which isn't full-screen.

Probably a better solution would be to convert the modal to a bottom sheet on mobile (as seen in other component libraries like [Catalyst](https://catalyst.tailwindui.com)).

While I do understand your proposed change, I hope you realize it doesn't expand to all other usages of modals. 🙂

**Reply by andrewdwallo** [reply]
> Imagine a modal that simply requests a name and email address, or even a modal with a single form field. Covering the entire page would hide all context for the user, and it would degrade the user experience compared to having a modal which isn't full-screen.

I mean I don't really think something like this:

<img width="1918" alt="Screenshot 2024-04-18 at 6 41 20 PM" src="https://github.com/filamentphp/filament/assets/104294090/26751813-4675-4c37-bcf3-2515251e2a3d">

Or this:
<img width="1918" alt="Screenshot 2024-04-18 at 6 42 53 PM" src="https://github.com/filamentphp/filament/assets/104294090/d4eda575-682b-4c12-9b31-e10373c38741">

Really degrades the user experience at all, compared to what it would look like currently:
<img width="1918" alt="Screenshot 2024-04-18 at 6 44 06 PM" src="https://github.com/filamentphp/filament/assets/104294090/7c2a8bcd-9fea-43a3-84b2-5c0199a082e2">

Both of them kind of hide the context of what is happening in the background anyway. Also, I don't think the user's focus would be on the background as much as it would the modal.

Another benefit of having it full screen on mobile is that if you have dynamic hidden/visible fields or select dropdowns, the modal won't be expanding in height possibly interrupting the user experience which would probably look worse on mobile.

Even then, a modal will have a Title and possibly a description which is already intended to give context of what is happening. 

If y'all won't consider making it full screen on mobile by default, then what about just a simple optional option to utilize the feature, using something such as this?:
```php
 use Filament\Support\Enums\MaxWidth;
 
Action::make('deleteBankConnection')
    ->form([
        // ...
    ])
    ->action(function (array $data): void {
        // ...
    })
    ->modalWidth(MaxWidth::TwoExtraLarge)
    ->fullWidthOnMobile() // or some better name
``` 

> While I do understand your proposed change, I hope you realize it doesn't expand to all other usages of modals.

While I appreciate you understanding and being polite about it, my original intention and proposed change was to never make it the default on mobile (it was suggested and I agreed with Dan though), but my original intention was to make it an optional feature, which means it doesn't have to cover all cases. I also believe the benefit of having the optional feature would be worth maintaining and having even though it doesn't cover just a few, probably rarer, cases as you mentioned...

**Reply by danharrin** [reply]
Ideally, we'd have something like

```
modalWidth([
    'default' => MaxWidth::Screen,
    'sm' => MaxWidth::TwoExtraLarge,
])
```

because people will prob want to customize the sm breakpoint

**Reply by zepfietje** [reply]
That could be a good solution to maintain BC indeed.

**Reply by andrewdwallo** [reply]
That would be great to have!

**Comment by andrewdwallo** [comment]
Is there any update on this? Is it still something that could be planned for maybe v4?

**Reply by danharrin** [reply]
I'd accept a PR based on my last comment for v3 or v4
